source/fuzz/transformation_add_parameter.cpp

```diff
@@ -24,40 +24,74 @@ TransformationAddParameter::TransformationAddParameter(
     : message_(message) {}
 
 TransformationAddParameter::TransformationAddParameter(
-    uint32_t function_id, uint32_t parameter_fresh_id, uint32_t initializer_id,
+    uint32_t function_id, uint32_t parameter_fresh_id,
+    uint32_t parameter_type_id, std::map<uint32_t, uint32_t> call_parameter_ids,
     uint32_t function_type_fresh_id) {
   message_.set_function_id(function_id);
   message_.set_parameter_fresh_id(parameter_fresh_id);
-  message_.set_initializer_id(initializer_id);
+  message_.set_parameter_type_id(parameter_type_id);
+  *message_.mutable_call_parameter_ids() =
+      fuzzerutil::MapToRepeatedUInt32Pair(call_parameter_ids);
   message_.set_function_type_fresh_id(function_type_fresh_id);
 }
 
 bool TransformationAddParameter::IsApplicable(
     opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
-  // Check that function exists
+  // Check that function exists.
   const auto* function =
       fuzzerutil::FindFunction(ir_context, message_.function_id());
   if (!function ||
       fuzzerutil::FunctionIsEntryPoint(ir_context, function->result_id())) {
     return false;
   }
 
-  // Check that |initializer_id| is valid.
-  const auto* initializer_inst =
-      ir_context->get_def_use_mgr()->GetDef(message_.initializer_id());
-
-  if (!initializer_inst) {
+  // The type must be supported.
+  uint32_t new_parameter_type_id = message_.parameter_type_id();
+  auto new_parameter_type =
+      ir_context->get_type_mgr()->GetType(new_parameter_type_id);
+  if (!new_parameter_type) {
     return false;
   }
-
-  // Check that initializer's type is valid.
-  const auto* initializer_type =
-      ir_context->get_type_mgr()->GetType(initializer_inst->type_id());
-
-  if (!initializer_type || !IsParameterTypeSupported(*initializer_type)) {
+  if (!IsParameterTypeSupported(*new_parameter_type)) {
     return false;
   }
 
+  // Iterate over all callers.
+  std::map<uint32_t, uint32_t> call_parameter_ids_map =
+      fuzzerutil::RepeatedUInt32PairToMap(message_.call_parameter_ids());
+  for (auto* instr :
+       fuzzerutil::GetCallers(ir_context, message_.function_id())) {
+    uint32_t caller_id = instr->result_id();
+
+    // If there is no entry for this caller, return false.
+    if (call_parameter_ids_map.find(caller_id) ==
+        call_parameter_ids_map.end()) {
+      return false;
+    }
+    uint32_t value_id = call_parameter_ids_map[caller_id];
+
+    auto value_instr = ir_context->get_def_use_mgr()->GetDef(value_id);
+    if (!value_instr) {
+      return false;
+    }
+    // If the id of the value of the map is not available before the caller,
+    // return false.
+    if (!fuzzerutil::IdIsAvailableBeforeInstruction(ir_context, instr,
+                                                    value_id)) {
+      return false;
+    }
```

Comment:
This can potentially trigger a bug if the caller is in an unreachable block. `fuzzerutil::IdIsAvailableBeforeInstruction` uses dominator analysis to check that `value_id` is available and the domination rules are not defined for unreachable blocks.

A solution is to adjust `fuzzerutil::IdIsAvailableBeforeInstruction` so that it uses DFS instead of the DominatorAnalysis. Or we could just check that the block is not unreachable.

Either way, I think it is better to leave it as it is for now and create a separate issue for this. @paulthomson @antnkarp , what do you think?

Reply:
I think this issue is common for many transformations. Maybe we could implement our own alternative to DominatorAnalysis or adjust the existing one?

Reply:
If it is common for many transformations, I agree that we need to create a sperate issue.

Reply:
Please, add a TODO(#3722) comment explaining that we need to refactor this. Use an original issue URL instead of #...